src/System.Net.Http/tests/FunctionalTests/ManagedHandlerTest.cs

```diff
@@ -391,5 +391,96 @@ public sealed class ManagedHandler_HttpClientHandler_ConnectionPooling_Test : Ht
                 }
             }
         }
+
+        [OuterLoop]
+        [Fact]
+        public async Task HttpConnectMethod_Succeeds()
+        {
+            using (HttpClient client = CreateHttpClient())
+            using (var listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+                listener.Listen(100);
+                var ep = (IPEndPoint)listener.LocalEndPoint;
+                var uri = new Uri($"http://{ep.Address}:{ep.Port}/");
+                string responseBody = "HTTP/1.1 200 Connection established\r\n\r\n";
+
+                HttpRequestMessage request = new HttpRequestMessage(new HttpMethod("CONNECT"), uri);
+                Task<HttpResponseMessage> responseTask = client.SendAsync(request);
+
+                using (Socket server = await listener.AcceptAsync())
+                using (var serverStream = new NetworkStream(server, ownsSocket: false))
+                {
+                    // Skip request headers.
+                    while (true)
+                    {
+                        if (serverStream.ReadByte() == '\r')
+                        {
+                            serverStream.ReadByte();
+                            break;
+                        }
+                        while (serverStream.ReadByte() != '\r') { }
+                        serverStream.ReadByte();
+                    }
+
+                    // Send response.
+                    await server.SendAsync(
+                    new ArraySegment<byte>(Encoding.ASCII.GetBytes(responseBody)),
+                                            SocketFlags.None);
+                    HttpResponseMessage response = await responseTask;
+                    Assert.Equal(response.StatusCode, HttpStatusCode.OK);
+                }
```

Comment:
We should do some validation here that we are actually able to send bits back and forth.  Minimally we should have the server send some data and have the client receive and verify it, and then vice-versa.

Reply:
Do you have more specific proposal @geoffkizer ??? I struggled with this as public API does not provide access to HttpConnection. With that I'm not sure what would be best way how to verify that connection can send random bits. 
My plan was to use it in combination with proxy and https request when that is ready.

Reply:
We should model this after the Upgrade stuff we already added for WebSockets.  Basically, return a RawConnectionStream, then access that via the response object.  (And use HttpCompletionOption as I mentioned above)

The WebSockets code that consumes the Upgrade stuff is here: https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs#L109

That said, I've just noticed that the WebSocket code is actually using HttpClientHandler, not HttpClient.  This is why it doesn't hit the buffering issue.

Unfortunately, HttpClientHandler.SendAsync is not public, so the WebSocket code is doing a semi-hack where it derives its own class from HttpClientHandler, so it can call HttpClientHandler.SendAsync.  For testing purposes, we could do the same here.

This all seems a little bit ugly, but I'm not sure what to do about it.  Adding @stephentoub to see if he has any thoughts.

BTW, if a user tries to use the Upgrade mechanism, I assume they will hit the same hang that you're seeing if they don't use HttpCompletionOption.  This seems bad too.

Reply:
Also, we should add HttpClient-level Upgrade tests.  I'll spend a little time on this later.

Reply:
> Adding @stephentoub to see if he has any thoughts.

1. I'm not particularly concerned about direct usage of CONNECT. It should be exceedingly rare. And to my knowledge none of the other handlers support it.
2. You can use the base HttpMessageInvoker class, which is a very thin wrapper around a handler, really just with a public SendAsync that calls the protected one. Or just specify the completion option to HttpClient that doesn't wait for the full response.

Reply:
Sounds reasonable to me.  I assume same goes for direct usage of Upgrade?  (i.e. we're fine if it hangs?)

Reply:
BTW, if we do want to fix the "hang" behavior, I think it's easy enough to do:  just test the response stream for CanRead and don't buffer in that case.

Reply:
I mean CanWrite.

Reply:
> I assume same goes for direct usage of Upgrade? 

Yes.  I'm not aware of a robust alternative.  For better or worse, the default is to buffer, and folks that don't want response buffering need to know to disable it.

Reply:
You don't think testing for CanWrite is robust?  I suppose someone (including us) could expose a response content stream that's writable, so yeah, maybe it's not ideal... 

Another idea would be to just explicitly do (stream is RawConnectionStream).  This is internal so it's a little weird, but maybe that's fine -- we don't really want any 3rd party to be able to turn off buffering.

Regardless, it's not a huge deal, I'm happy either way.